File: src/lib/digitizer/domain.ts
import domainDocument from "../../../config/ecg-domain.v1.json"

export const ECG_LEADS = [
  "I",
  "II",
  "III",
  "aVR",
  "aVL",
  "aVF",
  "V1",
  "V2",
  "V3",
  "V4",
  "V5",
  "V6",
] as const

export type EcgLead = (typeof ECG_LEADS)[number]
export type CanonicalEcgLayout = "standard_3x4" | "standard_6x2" | "standard_12x1"
export type PartialEcgLayout =
  | "standard_6x1_limb"
  | "cabrera_6x1_limb"
  | "precordial_6x1"
  | "precordial_3x2"
export type EcgLayout = CanonicalEcgLayout | PartialEcgLayout

export type DigitizerCandidateKind =
  | "source-model"
  | "adaptive-preprocessed"
  | "geometry-constrained"
  | "native-grid"
  | "crop-proposal"
  | "panel-component"
  | "panel-composite"
  | "peer-gap-repair"
  | "lead-fusion"

export type DigitizerCandidateCapabilities = {
  derived: boolean
  nativeGrid: boolean
  preprocessed: boolean
  geometryConstrained: boolean
  sourcePixelEvidence: boolean
  selectionEligible: boolean
}

export type EcgLayoutSpec = {
  pageRows: number
  rows: readonly (readonly string[])[]
}

function validateDomainDocument() {
  if (domainDocument.version !== 1) {
    throw new Error(`Unsupported ECG domain version: ${domainDocument.version}`)
  }
  if (JSON.stringify(domainDocument.leads) !== JSON.stringify(ECG_LEADS)) {
    throw new Error("ECG domain lead order does not match the canonical 12-lead order.")
  }
  for (const [layout, rawSpec] of Object.entries(domainDocument.layouts)) {
    if (!Number.isInteger(rawSpec.pageRows) || rawSpec.pageRows <= 0) {
      throw new Error(`ECG domain layout ${layout} has an invalid page row count.`)
    }
    const unknownLead = rawSpec.rows.flat().find(
      (lead) => !ECG_LEADS.includes(lead as EcgLead)
    )
    if (unknownLead) {
      throw new Error(`ECG domain layout ${layout} contains unknown lead ${unknownLead}.`)
    }
  }
}

validateDomainDocument()

export const SAMPLE_RATE_HZ = domainDocument.sampleRateHz
export const PAGE_DURATION_SECONDS = domainDocument.pageDurationSeconds
export const ECG_LAYOUTS: Readonly<Record<string, EcgLayoutSpec>> =
  Object.freeze(domainDocument.layouts)

export const PARTIAL_LAYOUT_LEADS: Readonly<Record<string, readonly string[]>> =
  Object.freeze({
    standard_6x1_limb: [
      ...new Set(ECG_LAYOUTS.standard_6x1_limb.rows.flat()),
    ],
    cabrera_6x1_limb: [
      ...new Set(ECG_LAYOUTS.cabrera_6x1_limb.rows.flat()),
    ],
    precordial_6x1: [...new Set(ECG_LAYOUTS.precordial_6x1.rows.flat())],
    precordial_3x2: [...new Set(ECG_LAYOUTS.precordial_3x2.rows.flat())],
  })

export const LAYOUT_COLUMNS: Readonly<
  Record<string, Readonly<Record<string, number>>>
> = Object.freeze(
  Object.fromEntries(
    Object.entries(ECG_LAYOUTS).map(([layout, spec]) => [
      layout,
      Object.fromEntries(
        spec.rows.flatMap((row) => row.map((lead, column) => [lead, column]))
      ),
    ])
  ) as Record<string, Record<string, number>>
)

File: ecg_pipeline/domain.py
from __future__ import annotations

import json
from pathlib import Path
from typing import Final


_DOMAIN_PATH = Path(__file__).resolve().parents[1] / "config" / "ecg-domain.v1.json"
_DOMAIN = json.loads(_DOMAIN_PATH.read_text(encoding="utf-8"))
if _DOMAIN.get("version") != 1:
    raise RuntimeError(f"Unsupported ECG domain version in {_DOMAIN_PATH}")

SAMPLE_RATE_HZ: Final[int] = int(_DOMAIN["sampleRateHz"])
PAGE_DURATION_SECONDS: Final[int] = int(_DOMAIN["pageDurationSeconds"])
LEAD_ORDER: Final[tuple[str, ...]] = tuple(_DOMAIN["leads"])
if LEAD_ORDER != (
    "I",
    "II",
    "III",
    "aVR",
    "aVL",
    "aVF",
    "V1",
    "V2",
    "V3",
    "V4",
    "V5",
    "V6",
):
    raise RuntimeError("The ECG domain does not contain the canonical lead order.")

LAYOUT_ROWS: Final[dict[str, tuple[tuple[str, ...], ...]]] = {
    name: tuple(tuple(row) for row in spec["rows"])
    for name, spec in _DOMAIN["layouts"].items()
}
SIX_BY_TWO_ROW_LEADS: Final[tuple[tuple[str, ...], ...]] = LAYOUT_ROWS[
    "standard_6x2"
]
THREE_BY_FOUR_ROW_LEADS: Final[tuple[tuple[str, ...], ...]] = LAYOUT_ROWS[
    "standard_3x4"
]
TWELVE_ROW_LEADS: Final[tuple[tuple[str, ...], ...]] = LAYOUT_ROWS[
    "standard_12x1"
]
SIX_BY_TWO_PANEL_SAMPLES: Final[int] = (
    SAMPLE_RATE_HZ * PAGE_DURATION_SECONDS // 2
)
THREE_BY_FOUR_PANEL_SAMPLES: Final[int] = (
    SAMPLE_RATE_HZ * PAGE_DURATION_SECONDS // 4
)
